onnxruntime/core/providers/cuda/nn/dropout.cc

```diff
@@ -41,5 +32,67 @@ ONNX_OPERATOR_KERNEL_EX(
         .InputMemoryType<OrtMemTypeCPUInput>(2),
     Dropout);
 
+Status Dropout::ComputeInternal(OpKernelContext* context) const {
+  //Get X_data
+  const Tensor* X = context->Input<Tensor>(0);
+  if (X == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "X Input is not available.");
+  const TensorShape& shape = X->Shape();
+  const int64_t N = shape.Size();
+
+  //Get Y_data
+  auto Y = context->Output(0, shape);
+
+  //Get mask_data
+  auto mask = context->Output(1, shape);
+  ORT_ENFORCE(!mask || mask->Shape().Size() == N);
+
+  //Get the ratio_data
+  float ratio_data = default_ratio_;
+  auto ratio = context->Input<Tensor>(1);
+  if (ratio) {
+    utils::MLTypeCallDispatcher<float, MLFloat16, double> t_disp(ratio->GetElementType());
+    t_disp.Invoke<GetRatioDataImpl>(ratio, ratio_data);
+  }
+
+  const Tensor* training_mode = context->Input<Tensor>(2);
+  //Check for inference mode.
+  if ((0 == ratio_data) ||(training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
+    const void* X_data = X->DataRaw();
+    void* Y_data = Y->MutableDataRaw();
+    if (Y_data != X_data) {
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(Y_data, X_data, X->SizeInBytes(), cudaMemcpyDeviceToDevice, Stream()));
+    }
+
+    // If mask is requested, return all 1s.
+    if (mask != nullptr) {
+      CUDA_RETURN_IF_ERROR(cudaMemsetAsync(mask->MutableData<bool>(), true, N * sizeof(bool), Stream()));
+    }
+
+    return Status::OK();
+  }
+
+  IAllocatorUniquePtr<bool> temp_mask_buffer{};  // buffer to use if mask is not provided
+  bool* const mask_data = [this, N, mask, &temp_mask_buffer]() {
+    if (mask) return mask->MutableData<bool>();
+    temp_mask_buffer = GetScratchBuffer<bool>(N);
+    return temp_mask_buffer.get();
+  }();
+
+  PhiloxGenerator& generator = generator_ ? *generator_ : PhiloxGenerator::Default();
+
+  using SupportedTypes = onnxruntime::TypeList<
```

Comment:
>SupportedTypes [](start = 8, length = 14)

this could be consolidated with the ALL_FLOAT_X_TYPES macros

Reply:
good catch. done.